### PR TITLE
Simplify detect preview feature and fix NRE

### LINKED ISSUE #5802: CSharpDetectPreviewFeatureAnalyzer hitting NullReferenceException when run on Corelib
### Analyzer

**Diagnostic ID**: [CA2252](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2252): `Opt in to preview features`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 6.0.3 (Latest)

### Describe the bug

e.g. AD0001 fired with the exception from the analyzer:
```
Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Message	AD0001	Analyzer 'Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.
Exception occurred with following context:
Compilation: System.Private.CoreLib
ISymbol: Decimal (NamedType)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer.GetPreviewInterfaceNodeForTypeImplementingPreviewInterface(ISymbol typeSymbol, ISymbol previewInterfaceSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.ProcessTypeSymbolAttributes(SymbolAnalysisContext context, ITypeSymbol symbol, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, IFieldSymbol virtualStaticsInInterfaces, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.<>c__DisplayClass33_0.<Initialize>b__2(SymbolAnalysisContext context)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c.<ExecuteSymbolActionsCore>b__50_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----

Suppress the following diagnostics to disable this analyzer: CA2252	System.Private.CoreLib		1	Active	Analyzer 'Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer' threw the following exception:
'Exception occurred with following context:
Compilation: System.Private.CoreLib
ISymbol: Decimal (NamedType)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer.GetPreviewInterfaceNodeForTypeImplementingPreviewInterface(ISymbol typeSymbol, ISymbol previewInterfaceSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.ProcessTypeSymbolAttributes(SymbolAnalysisContext context, ITypeSymbol symbol, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, IFieldSymbol virtualStaticsInInterfaces, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.<>c__DisplayClass33_0.<Initialize>b__2(SymbolAnalysisContext context)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c.<ExecuteSymbolActionsCore>b__50_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----
'.
```
cc: @jeffhandley

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #5802

@jeffhandley @buyaa-n The bit of refactoring I made have caused some changes in diagnostic locations.

The new code is more reliable as it compares actual symbols instead of dealing with syntax nodes which can give inaccurate results.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs [hunk 1]
```diff
@@ -212,7 +212,28 @@ public abstract class DetectPreviewFeatureAnalyzer : DiagnosticAnalyzer
...
+                INamedTypeSymbol typeSymbol => typeSymbol.TypeParameters,
+                IMethodSymbol methodSymbol => methodSymbol.TypeParameters,
+                _ => throw new ArgumentException($"Unexpected symbol '{typeOrMethodSymbol.Kind}'")
+            };
+
+            foreach (var typeParameter in typeParameters)
+            {
+                foreach (var constraintType in typeParameter.ConstraintTypes)
+                {
+                    if (previewInterfaceConstraintSymbol.Equals(constraintType, SymbolEqualityComparer.Default /*DetectPreviewFeatureSymbolEqualityComparer*/))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:228]: PR isn't ready to merge until I make sure symbol comparison will behave the way it should, potentially by introducing a new comparer.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDetectPreviewFeatureAnalyzer.cs [hunk 2]
```diff
@@ -241,25 +179,8 @@ private static bool TryGetConstraintClauseNode(SyntaxList<TypeParameterConstrain
...
-                else if (typeSymbolDefinition is StructDeclarationSyntax structDeclaration)
-                {
-                    SeparatedSyntaxList<BaseTypeSyntax> baseListTypes = structDeclaration.BaseList.Types;
-                    if (TryGetPreviewInterfaceNodeForClassOrStructImplementingPreviewInterface(baseListTypes, previewInterfaceSymbol, out ret))
-                    {
-                        return ret;
-                    }
-                }
-                else if (typeSymbolDefinition is InterfaceDeclarationSyntax interfaceDeclaration)
+                if (typeSymbolDefinition is TypeDeclarationSyntax { BaseList.Types: var baseListTypes })
```
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDetectPreviewFeatureAnalyzer.cs:182]: This change fixes two issues:

- The null reference exception
- record declarations

**Comment by jeffhandley** [comment]
Thanks for working on this, @Youssef1313. Those diagnostic locations were chosen intentionally though, and if I understand the changes correctly from a quick scan, I don't think we'd want to take the changes that move the diagnostic locations.

Here are some of the PRs that affected the locations before:
* https://github.com/dotnet/roslyn-analyzers/pull/5402
* https://github.com/dotnet/roslyn-analyzers/pull/5425

The strategy on these was to put the diagnostic on the preview API that is referred to, as opposed to the API that is referencing the preview API. Does that make sense? One acceptance criterion there was that in dotnet/runtime, we needed to be able to have code like this: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Int32.cs#L13-L24

```cs
    [Serializable]
    [StructLayout(LayoutKind.Sequential)]
    [TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
    public readonly struct Int32 : IComparable, IConvertible, ISpanFormattable, IComparable<int>, IEquatable<int>
#if FEATURE_GENERIC_MATH
#pragma warning disable SA1001, CA2252 // SA1001: Comma positioning; CA2252: Preview Features
        , IBinaryInteger<int>,
          IMinMaxValue<int>,
          ISignedNumber<int>
#pragma warning restore SA1001, CA2252
#endif // FEATURE_GENERIC_MATH
    {
```

In that code, we need to limit the `#pragma warning disable` to be only around the interfaces that are in preview, without having that `#pragma warning disable` wrapped around the `Int32` type declaration itself. This guarantees that the suppressions are specific to intentional references, without inadvertently over-suppressing.

Note that moving a diagnostic location is a subtle breaking change, and we actually changed those diagnostic locations in .NET 6.0 RC2 as a breaking change between RC1 and RC2 (https://github.com/dotnet/runtime/issues/57224#issuecomment-899856534). When we made those changes, we had to introduce the syntax-specific code paths since there aren't any syntax-agnostic APIs to get us the desired diagnostic locations.

**Comment by Youssef1313** [comment]
Thanks for the context @jeffhandley 
One point to note is diagnostic location is not on the type declaration. Here is an example diff:

image

But I see other concerns are still applicable.

**Pull request closed without merging**